text/0625-helper-managers.md

```diff
@@ -0,0 +1,861 @@
+}
+
+type HelperCapabilities = Opaque;
+
+export declare function capabilities(
+  // to be replaced with the version of Ember this lands in
+  version: '3.21.0',
+  options: HelperCapabilitiesOptions
+): HelperCapabilities;
+```
+
+Helper managers themselves have the following interface:
+
+```ts
+interface HelperManager<HelperStateBucket> {
+  capabilities: HelperCapabilities;
+
+  createHelper(definition: HelperDefinition, args: TemplateArgs): HelperStateBucket;
+
+  getValue?(bucket: HelperStateBucket, args: TemplateArgs): unknown;
+
+  runEffect?(bucket: HelperStateBucket, args: TemplateArgs): void;
+
+  getDestroyable?(bucket: HelperStateBucket): object;
+}
+```
+
+Let's dig into these hooks one by one:
+
+### Hooks
+
+#### `createHelper`
+
+`createHelper` is a required hook on the HelperManager interface. The hook is
+passed the definition of the helper that is currently being created, and is
+expected to return a _state bucket_. This state bucket is what represents the
+current state of the helper, and will be passed to the other lifecycle hooks at
+appropriate times. It is not necessarily related to the definition of the
+helper itself - for instance, you could return an object _containing_ an
+instance of the helper:
+
+```js
+class MyManager {
+  createHelper(Definition, args) {
+    return {
+      instance: new Definition(args);
+    };
+  }
+}
+```
+
+This allows the manager to store metadata that it doesn't want to expose to the
+user.
+
+This hook is _not_ autotracked - changes to tracked values used within this hook
+will _not_ result in a call to any of the other lifecycle hooks. If users do
+want to autotrack some values used during construction, they can either create
+the instance of the helper in `runEffect` or `getValue`, or they can use the
+`cache` API to autotrack the `createHelper` hook themselves. This provides
+maximum flexibility and expressiveness to manager authors.
+
+This hook has the following timing semantics:
+
+**Always**
+- called as discovered during DOM construction
+- called in definition order in the template
+
+#### `getValue`
+
+`getValue` is an optional hook that should return the value of the helper. This
+is the value that is returned from the helper and passed into the template.
+
+This hook is called when the value is requested from the helper (e.g. when the
+template is rendering and the helper value is needed). The hook is autotracked,
+and will rerun whenever any tracked values used inside of it are updated.
+Otherwise it does not rerun.
+
+> Note: This means that arguments which are not _consumed_ within the hook will
+> not trigger updates.
+
+This hook is only called for helpers with the `hasValue` capability enabled.
+This hook has the following timing semantics:
+
+**Always**
+- called the first time the helper value is requested
+- called after autotracked state has changed
+
+**Never**
+- called if the `hasValue` capability is disabled
+
+#### `runEffect`
```

Comment:
Can you add some prose to clarify what the behavior is in SSR / FastBoot? As you know we do not run modifiers in SSR, and in general SSR does not have the ability to run update opcodes. I could easily imagine that we say `runEffect` should **not** run in SSR (essentially saying that you shouldn't have external side effects), but we should be clear and we should consider it carefully.

Reply:
I think we _probably_ should say they don't run in SSR for now. I can absolutely imagine wanting them eventually in SSR, but I think the safe default to start with (especially while introducing this new concept) is to disallow it. In the future, we can add a capability to explicitly declare that you do or do not want to run in SSR which means this isn't really blocking future functionality.